src/lib/y2storage/proposal/space_maker.rb

```diff
@@ -237,57 +269,81 @@ def resize_and_delete(planned_partitions, keep, lvm_helper, disk: nil)
-          [:resize_windows!,     [{ force: true }]],
-          # step 5 - delete Windows partitions
-          [:delete_partitions!,  [:windows, keep]],
-          # step 6 - if deleting partitions was not enough, maybe there is no
-          # partition table and we have to wipe the disk
-          [:delete_disk_content, [lvm_helper]]
-        ]
-
-        methods_and_args.each_with_index do |method_and_args, idx|
-          method, extra_args = method_and_args
-
-          log.info "Step #{idx + 1} - #{method}"
-          send(method, planned_partitions, disk, *extra_args)
-
-          break if @distribution
+        @disk_name = disk
+        calculate_sorted_candidates(keep)
+
+        until success?(planned_partitions)
+          break unless execute_next_action(planned_partitions, lvm_helper)
         end
 
         raise Error unless @distribution
       end
 
+      # Performs the next action of {#resize_and_delete}
+      #
+      # @return [Boolean] true if some operation was performed. False if nothing
+      #   else could be done to reach the goal.
+      def execute_next_action(planned_partitions, lvm_helper)
+        part = next_partition_to_resize
+        if part
+          target_shrink_size = resizing_size(part, planned_partitions)
+          shrink_partition(part, target_shrink_size)
+        else
+          part_sid = next_partition_to_delete
+          if part_sid
+            # Strictly speaking, this could lead to deletion of a partition
+            # included in the keep array. In practice it doesn't matter because
+            # PVs and extended partitions are never marked to be reused as a
+            # planned partition.
+            sids = new_graph_part_killer.delete_by_sid(part_sid)
+            new_graph_deleted_sids.concat(sids)
+          else
+            wipe = next_disk_to_wipe(lvm_helper)
+            return false unless wipe
+            remove_content(wipe)
+          end
+        end
+
+        true
+      end
+
+      # Fills up {#sorted_candidates} for the current execution of
+      # {#resize_and_delete}
+      def calculate_sorted_candidates(keep)
+        @sorted_candidates = {}
+        [:linux, :other, :windows].each do |type|
+          @sorted_candidates[type] =
+            if settings.delete_forbidden?(type)
+              []
+            else
+              deletion_candidate_partitions(new_graph, type, disk_name) - keep
+            end
+        end
+      end
+
+      # Plain list of all candidates to be deleted
+      #
+      # @return [Array<Integer>]
+      def all_sorted_candidates
+        sorted_candidates[:linux] + sorted_candidates[:other] + sorted_candidates[:windows]
+      end
+
+      # Sid of the next partition to be deleted by {#resize_and_delete}
+      #
+      # @return [Integer]
+      def next_partition_to_delete
+        result = (all_sorted_candidates - new_graph_deleted_sids).first
+
+        log.info "Next partition to delete: #{result}"
```

Comment:
NP: it is a bit weird to see in log: "Next partition to delete: nil". I would print a more specific message when there is no partition to delete.

Reply:
Actually I find it kind of convenient to have a stable string for grepping.